ngraph/core/src/op/gather_nd.cpp

```diff
@@ -17,11 +17,158 @@
+    : Op({ data, indices })
+    , m_batch_dims(batch_dims)
+{
+    constructor_validate_and_infer_types();
+}
+
+void op::v5::GatherND::validate_and_infer_types()
+{
+    // check types of input tensors
+    const auto& data_type = get_input_element_type(0);
+    const auto& indices_type = get_input_element_type(1);
+
+    NODE_VALIDATION_CHECK(this,
+        indices_type.is_integral_number(),
+        "The indices type is expected to be an integer type. Got: ",
+        indices_type);
+
+    // check ranks of input tensors
+    const auto& data_pshape = get_input_partial_shape(0);
+    const auto& indices_pshape = get_input_partial_shape(1);
+
+    if (data_pshape.rank().is_static()) {
+        NODE_VALIDATION_CHECK(this,
+            data_pshape.rank().get_length() > 0,
+            "Data rank must be at least 1.");
+
+        NODE_VALIDATION_CHECK(this,
+            data_pshape.rank().get_length() > m_batch_dims,
+            "Number of batch dimensions must not exceed a rank of data.");
+    }
+
+    if (indices_pshape.rank().is_static()) {
+        NODE_VALIDATION_CHECK(this,
+            indices_pshape.rank().get_length() > 0,
+            "Indices rank must be at least 1.");
+
+        NODE_VALIDATION_CHECK(this,
+            indices_pshape.rank().get_length() > m_batch_dims,
+            "Number of batch dimensions must not exceed a rank of indices.");
+    }
+
+    if (data_pshape.rank().is_static() && indices_pshape.rank().is_static()) {
+        // check that batch dimensions of data and indices are the same
+        for (auto batch_dim = 0; batch_dim < m_batch_dims; batch_dim++) {
+            if (data_pshape[batch_dim].is_static() && indices_pshape[batch_dim].is_static()) {
+                NODE_VALIDATION_CHECK(this,
+                    data_pshape[batch_dim].get_length() == indices_pshape[batch_dim].get_length(),
+                    "Batch dimensions of data and indices must be the same.");
+            }
+        }
+
+        if (indices_pshape[indices_pshape.rank().get_length() - 1].is_static()) {
+            NODE_VALIDATION_CHECK(this,
+                (indices_pshape[indices_pshape.rank().get_length() - 1].get_length() + m_batch_dims) <= data_pshape.rank().get_length(),
+                "Length of a tuple with indices must not exceed a rank of data tensor excluding batch dimensions.");
+        }
+    }
+
+    // set output shape
+    set_output_size(1);
+    if (data_pshape.rank().is_static() && indices_pshape.rank().is_static() &&
+        indices_pshape[indices_pshape.rank().get_length() - 1].is_static()) {
+        auto indices_tuple_length = indices_pshape[indices_pshape.rank().get_length() - 1].get_length();
+        auto slice_length = data_pshape.rank().get_length() - indices_tuple_length - m_batch_dims;
+        auto output_indices_length = indices_pshape.rank().get_length() - m_batch_dims - 1;
+        if (m_batch_dims > 0)
+        {
+            auto output_rank = 1 + output_indices_length + slice_length;
```

Comment:
Looks like we have a lot of code duplication in those if/else branches.
We can reuse some code:
```
auto output_rank = output_indices_length + slice_length;
std::vector<Dimension> output_shape(output_rank + 1);
int shift = 1;
if (m_batch_dims > 0) {
    for (auto dim = 0; dim < m_batch_dims; dim++) {
    ...
} else {
    shift = 0;
    output_shape.pop_back();
}

for (auto dim = 0; dim < output_indices_length; dim++)
{
     output_shape[dim + shift] = indices_pshape[dim + m_batch_dims];
}
for (auto dim = 0; dim < slice_length; dim++)
{
    output_shape[output_indices_length + dim + shift] = data_pshape[m_batch_dims + indices_tuple_length + dim];
}
set_output_type(0, data_type, PartialShape(output_shape));
```

Reply:
Done. Removed the duplication.